dkc/core/models/tree.py

```diff
@@ -15,3 +19,23 @@ def root_folder(self) -> Folder:
         from .folder import Folder
 
         return Folder.objects.get(tree=self, parent=None)
+
+    @classmethod
+    def filter_by_permission(
+        cls, user: User, permission: Permission, queryset: models.QuerySet['Tree']
+    ) -> models.QuerySet['Tree']:
+        perms = permission.derived_permissions
+        return get_objects_for_user(user, perms, klass=queryset, any_perm=True)
+
+    def has_permission(self, user: User, permission: Permission) -> bool:
+        for perm in permission.derived_permissions:
+            if user.has_perm(perm, self):
+                return True
+        return False
```

Comment:
I think this is more readable / expressive as

```python
return any(user.has_perm(perm, self) for perm in permission.derived_permissions)
```

Also I find the name "derived_permissions" confusing here, as reading it out of context makes it seem backwards to what you expect. It's only the derived permissions if you're talking about *required* permission, not the permission *assigned* to a user. I'm struggling to come up with a great name, but I think capturing that it only works in that direction would be prudent.